extensions-core/druid-basic-security/src/main/java/org/apache/druid/security/basic/authentication/validator/PasswordHashGenerator.java

```diff
@@ -0,0 +1,156 @@
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Generates a hash of passwords using the {@link #HASH_ALGORITHM}. Multiple
+ * iterations may be used to enhance security of the generated hash.
+ * <p>
+ * Hashes once computed are cached for an hour so that they need not be recomputed
+ * for every API invocation.
+ */
+public class PasswordHashGenerator
+{
+  private static final Logger log = new Logger(PasswordHashGenerator.class);
+
+  public static final int KEY_LENGTH = 512;
+  public static final String HASH_ALGORITHM = "PBKDF2WithHmacSHA512";
+
+  /**
+   * Salt used to compute a quick sha-256 hash of the password for the cache.
+   */
+  private final byte[] shaSalt = BasicAuthUtils.generateSalt();
+
+  private final Cache<CacheKey, byte[]> cache = CacheBuilder.newBuilder()
+                                                            .maximumSize(1000)
+                                                            .expireAfterAccess(Duration.ofMinutes(60))
+                                                            .build();
+
+  /**
+   * Hashes the given password using the {@link #HASH_ALGORITHM}.
+   */
+  public byte[] getOrComputePasswordHash(char[] password, byte[] salt, int numIterations)
+  {
+    try {
+      return cache.get(
+          CacheKey.of(password, salt, numIterations, shaSalt),
+          () -> computePasswordHash(password, salt, numIterations)
+      );
+    }
+    catch (ExecutionException e) {
+      throw DruidException.defensive().build(e, "Could not compute hash of password");
+    }
+  }
+
+  /**
+   * Utility method to compuate hash of the given password using the {@link #HASH_ALGORITHM}.
+   * Callers should use the non-static method instead to leverage caching.
+   */
+  public static byte[] computePasswordHash(final char[] password, final byte[] salt, final int iterations)
+  {
+    try {
+      SecretKeyFactory keyFactory = SecretKeyFactory.getInstance(HASH_ALGORITHM);
+      SecretKey key = keyFactory.generateSecret(new PBEKeySpec(password, salt, iterations, KEY_LENGTH));
+      return key.getEncoded();
+    }
+    catch (InvalidKeySpecException ikse) {
+      log.error("Invalid keyspec");
+      throw new RuntimeException("Invalid keyspec", ikse);
+    }
+    catch (NoSuchAlgorithmException nsae) {
+      log.error("Hash algorithm[%s] is not supported on this system.", HASH_ALGORITHM);
+      throw new RE(nsae, "Hash algorithm[%s] is not supported on this system.", HASH_ALGORITHM);
+    }
+  }
+
+  /**
+   * Key used in the {@link #cache}. An SHA-256 hash of the password is used
+   * instead of the actual string so that the passwords are never exposed, even
+   * in a heap dump.
+   */
+  private static class CacheKey
+  {
+    final byte[] passwordSha;
+    final byte[] salt;
+    final int numIterations;
```

Comment:
Yes, that is because the final hash value can change if numIterations changes. Also, it doesn't seem like we use the same number of iterations for every credential as it is configurable. But let me double check to be sure.